Topic dependency graph input in the classroom

Nik-09 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->
The PR is stacked on top of https://github.com/oppia/oppia/pull/15953. This PR majorly includes graph input functionality i.e., adding topic ID to prerequisite topic IDs for creating topics dependency.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct


https://user-images.githubusercontent.com/57531197/193397114-75ef1ab2-50d3-444b-b89b-90f954c87611.mp4

**Updated video**

https://user-images.githubusercontent.com/57531197/193463522-82639746-5be9-4bdc-8e49-f39f6f7c0e01.mp4




## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

Nik-09 committed:
Adds diagnostic test model.

Nik-09 committed:
Created Classroom model.

Nik-09 committed:
Adds backend test into the shard.

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Merge branch 'classroom_model' of https://github.com/Nik-09/oppia into classroom_model

Nik-09 committed:
Addressed review comments.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_model

Nik-09 committed:
Adds backend coverage tests

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Fixes backend test.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_model

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_model

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_model

Nik-09 committed:
NMerge branch 'develop' of https://github.com/oppia/oppia into classroom_model

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_model

Nik-09 committed:
Adds domain object for classroom model.

Nik-09 committed:
Fixes lint issue.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_model

Nik-09 committed:
fixes lint issue

Nik-09 committed:
Merge branch 'classroom_model' of https://github.com/Nik-09/oppia into domain_object_for_classroom_model

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into domain_object_for_classroom_model

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Adds more backend unit-tests

Nik-09 committed:
Adds handler class for the classroom admin page

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into domain_object_for_classroom_model

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into domain_object_for_classroom_model

Nik-09 committed:
Merge branch 'domain_object_for_classroom_model' of https://github.com/Nik-09/oppia into controller_layer_addition_for_classroom

Nik-09 committed:
Adds classroom config services file

Nik-09 committed:
Adds test case structure

Nik-09 committed:
Adds test for the classroom config service

Nik-09 committed:
Adds backend unit tests for classroom admin handlers.

Nik-09 committed:
Fixes mypy, linter, backend unit tests.

Nik-09 committed:
Fixes lint issues.

Nik-09 committed:
Fixes merge conflicts

Nik-09 committed:
Adds backend unit test.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into controller_and_service_changes_for_classroom

Nik-09 committed:
Merge branch 'controller_and_service_changes_for_classroom' of https://github.com/Nik-09/oppia into backend_api_changes_for_classroom_admin

Nik-09 committed:
Adds backend api for the classroom admin.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into controller_and_service_changes_for_classroom

Nik-09 committed:
Addressed review comments.

Nik-09 committed:
Merge branch 'controller_and_service_changes_for_classroom' of https://github.com/Nik-09/oppia into backend_api_changes_for_classroom_admin

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into backend_api_changes_for_classroom_admin

Nik-09 committed:
Fixes lint issues.

Nik-09 committed:
Created diecrtory srructure for the classroom admin UI.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into backend_api_changes_for_classroom_admin

Nik-09 committed:
Adds admin feature to generate dummy classroom data

Nik-09 committed:
Add extra newlines for readability

Nik-09 committed:
Makes UI more interactive.

Nik-09 committed:
Adds modals for confirmation in various events

Nik-09 committed:
Merge branch 'develop' into backend_api_changes_for_classroom_admin

Nik-09 committed:
Refactored code for increasing readability.

Nik-09 committed:
Fixes merge conflict

Nik-09 committed:
Adds copyright comments in the starting of newly created files.

Nik-09 committed:
Adds textarea instead of input

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into backend_api_changes_for_classroom_admin

Nik-09 committed:
Addressed review comments.

Nik-09 committed:
Merge branch 'backend_api_changes_for_classroom_admin' of https://github.com/Nik-09/oppia into backend_api_changes_for_classroom_admin

Nik-09 committed:
Merge branch 'backend_api_changes_for_classroom_admin' of https://github.com/Nik-09/oppia into classroom_admin_ui

Nik-09 committed:
Adds backend test for admin

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Fixes lint issues.

Nik-09 committed:
Adds frontend unit test

Nik-09 committed:
Adds classroom admin component tests.

Nik-09 committed:
Adds frontend unit tests

Nik-09 committed:
UI changes.

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Adds backend unit test coverage

Nik-09 committed:
Fixes typescript checks in strict mode.

Nik-09 committed:
Merge branch 'develop' into classroom_admin_ui

Nik-09 committed:
Rephrases test descriptions.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Merge branch 'classroom_admin_ui' of https://github.com/Nik-09/oppia into classroom_admin_ui

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Removes dead code

Nik-09 committed:
Rephrases html text.

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Converts topic ID to topic Name.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Updates html text.

Nik-09 committed:
Rephrases test descriptions

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Adds topic dependency input feature

Nik-09 committed:
Fixes merge conflict

Nik-09 committed:
Fixes linter issues

Nik-09 committed:
Fixes mypy issues

Nik-09 committed:
Fixes typescript check

Nik-09 reviewed (commented):
Good morning @seanlip, I have created this PR for providing functionality to add topics dependency in a classroom.
The part to visually present the graph is still pending (I am thinking to refactor some of the current code and then present, but that changes is independent of the current changes)

Please provide your feedback on this PR i.e., stacked on top of https://github.com/oppia/oppia/pull/15953 and majorly focuses on adding/modifying the topics dependency.

Thank you

seanlip reviewed (changes requested):
Thanks, @Nik-09! Some thoughts about the UI: 

- "Classroom config properties" --> "Classroom details"
- Delete "Configure the details of a classroom"
- Delete "Existing Classrooms" (unless there is another option besides that? In which case, keep it. But can you show what that case looks like?)
- Topics and its prerequisites --> Topics and their prerequisites
- Change the subheading below that to just "The prerequisite topics below are used to generate the diagnostic test.", and italicize it.
- It is odd to have "Select Prerequisite" in the "Dummy Topic 1" column. Please keep that within the actual prerequisite list "column".
- The whole prerequisites UI is actually a bit odd and not that intuitive because the dropdown is a repeat of the prerequisites displayed. We talked about using select2 to address this. Is there any reason you didn't do so?
- It should be possible to reorder topics within the list.
- If you click the bottom right "Cancel" button after making a bunch of edits, what happens? Do all those edits disappear?

Nik-09 committed:
UI fixes

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Merge branch 'classroom_admin_ui' of https://github.com/Nik-09/oppia into topic_dependency_graph

Nik-09 committed:
Addressed review comments

Nik-09 commented:
> "Classroom config properties" --> "Classroom details"

Removed

> Delete "Configure the details of a classroom"

Removed

> Delete "Existing Classrooms" (unless there is another option besides that? In which case, keep it. But can you show what that case looks like?)

Removed

> Topics and its prerequisites --> Topics and their prerequisites

Updated

> Change the subheading below that to just "The prerequisite topics below are used to generate the diagnostic test.", and italicize it.

Updated

> It is odd to have "Select Prerequisite" in the "Dummy Topic 1" column. Please keep that within the actual prerequisite list "column".

Updated

> The whole prerequisites UI is actually a bit odd and not that intuitive because the dropdown is a repeat of the prerequisites displayed. We talked about using select2 to address this. Is there any reason you didn't do so?

“....the dropdown is a repeat of the prerequisites displayed” Please help me to understand this line.

> It should be possible to reorder topics within the list.

Please confirm, reorder the topics within the list means reordering the prerequisites for a topic. If yes, I think reordering is not essential because:
- The dependencies among the topic will be used by the diagnostic test algorithm for suggesting a topic to a learner. And the diagnostic test algorithm will have its own strategy to select a topic ID from the eligible set of topic IDs. So please help me to understand how ordering here will help the algorithm to perform better.
- Reference to the doc, [how topic ID is selected](https://docs.google.com/document/d/1VCl1JtF6ikgN1Ux4UDvkC9ubJdHnOOizu54ubh4vHok/edit#bookmark=id.6xha8bnv3ukp).
Also, I am assuming this in terms of reordering the prerequisite, please correct me if the above statement is referring to something else.


> If you click the bottom right "Cancel" button after making a bunch of edits, what happens? Do all those edits disappear?

After editing if one tries to click the cancel button then a modal will be presented asking for confirmation.

Nik-09 committed:
Addressed review comments

seanlip commented:
> “....the dropdown is a repeat of the prerequisites displayed” Please help me to understand this line.

The dropdown shows the same info displayed in the other field in a different form, i.e. a full list of prereqs with some of them ticked. But when adding a new topic I would expect to see only the list of things I can still add. I think the UI should support:
- add a new topic to the prereq topic list
- delete an existing topic from topic list

> It should be possible to reorder topics within the list.

To clarify, I don't mean reordering the prerequisites for a topic. I mean reordering the topics themselves because that's the order that would be shown in the classroom page, right?

>> If you click the bottom right "Cancel" button after making a bunch of edits, what happens? Do all those edits disappear?

> After editing if one tries to click the cancel button then a modal will be presented asking for confirmation.

Ah ok, thanks. And what does the modal say and what happens if the user confirms? (A video might help.)

Nik-09 committed:
Fixes merge conflict

Nik-09 committed:
Merge branch 'classroom_admin_ui' of https://github.com/Nik-09/oppia into topic_dependency_graph

Nik-09 committed:
Updates UI.

Nik-09 committed:
Fixes lint issues

Nik-09 reviewed (commented):
Hi, @seanlip I have updated the video for the topic and prerequisite addition/deletion option.
Please provide your feedback.

Also, I will shortly update the reordering of topics in a classroom but before that, I wanted to present and take feedback on the UI for the addition and deletion of topics.

Thank you.

seanlip commented:
Hi @Nik-09, I took a look at the video, and I think my main feedback is the same as I mentioned during our meeting -- it might be a good idea to explore select2 which might lead to a more consistent UI and not require extra special-case handling. Take a look and let me know what you think?

Thanks!

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Removes duplicate naming for dummy topics

Nik-09 committed:
Refactored classroom component as per review comments

Nik-09 committed:
Adds frontend unit test.

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Drag and drop + select2 dropdown

Nik-09 committed:
Fixes merge conflict

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Fixes UI issues.

Nik-09 reviewed (commented):
Hi, @seanlip I have updated the video, please take a look & provide your feedback on it.
Thank you.

seanlip commented:
Hi @Nik-09, I'm still a bit confused. Is there a reason you haven't tried using select2 yet?

For the dragging, I suggest: (a) having a "drag handler" on the left -- that's the thing that should be dragged, see the answer groups or rules editor in the exploration editor for an example, and (b) while dragging, don't show the thing that's being dragged on the initial list. Try to follow exactly the existing behaviour for where this sort of UI is present in the exploration editor.

Nik-09 committed:
Addressed review comments

Nik-09 committed:
Fixes lint issues

Nik-09 committed:
Improves UI for the topic dependency

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into classroom_admin_ui

Nik-09 committed:
Addressed review comments.

Nik-09 committed:
Adds graph visualization UI

Nik-09 committed:
Adds graph vizualization

Nik-09 committed:
Addressed review comments.

Nik-09 committed:
Adds tests for new classroom model

Nik-09 committed:
Fixes frontend tests

Nik-09 committed:
Fixes merge conflict

Nik-09 committed:
UI improvements

Nik-09 committed:
Fixes merge conflicts

Nik-09 committed:
Fixes lint issues.

Nik-09 reviewed (commented):
Good afternoon @seanlip and @kevintab95,
I have updated the UI and proof video as per the discussion.
Also, I am working on frontend tests but I need a review because of urgency. I will fix the tests in the next commit.

Please provide your feedback, once you are free.

Thank you.

Nik-09 committed:
Renames css class

seanlip reviewed (changes requested):
Thanks, @Nik-09! The UI looks great and the 2 codeowner files I have look good. The only comment I have on the UI is: in the error message about dependencies, please print the actual chain of dependencies and not just an example with "A" and "B". Say something like "topic 1 depends on topic 2 which depends on topic 1" etc. 

Make sure you have tests checking that the right message is generated for a range of different cases.

Otherwise, looks great, thanks!

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.html:
```diff
@@ -158,6 +273,88 @@ <h3>Classroom config properties</h3>
+    padding: 10px;
+  }
+  .drag-and-drop-list {
+    background: white;
```
Please prefer hex / rgb for specifying colours. See style guide: https://github.com/oppia/oppia/wiki/Coding-style-guide#css.

kevintab95 replied:
Ditto below.

Nik-09 replied:
Updated

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -42,36 +59,82 @@ export class ClassroomAdminPageComponent implements OnInit {
+  classroomIdToClassroomName: {[classroomId: string]: string} = {};
   existingClassroomNames: string[] = [];
 
+  myControl = new FormControl('');
```
What does myControl mean?

Nik-09 replied:
Updated with topicsFilter

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -42,36 +59,82 @@ export class ClassroomAdminPageComponent implements OnInit {
+  eligibleTopicNamesForPrerequisites!: string[];
+
+  topicIds: string[] = [];
+  eligibleTopicNames: string[] = [];
```
What are the topic names eligible for? I noticed that we have `eligibleTopicNamesForPrerequisites`, should update `eligibleTopicNames` to also have "for" part?

Nik-09 replied:
Updated

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -252,6 +329,210 @@ export class ClassroomAdminPageComponent implements OnInit {
     this.getAllClassroomIdToClassroomName();
   }
+
+  getTopicDependencyByTopicName(
```
The function seems to imply it is a getter, i.e. it returns something but really it is just setting a few things. Should we change it to  `setTopicDependencyByTopicName` instead?

Nik-09 replied:
Updated

kevintab95 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -42,36 +59,82 @@ export class ClassroomAdminPageComponent implements OnInit {
   classroomEditorMode: boolean = false;
   classroomDataSaveInProgress: boolean = false;
 
+  addNewTopicInputIsShown: boolean = false;
```
In general, avoid starting a variable name with a verb. Can we rename this to "newTopicCanBeAdded" or similar?

Nik-09 replied:
Updated

kevintab95 reviewed (changes requested):
Thanks @Nik-09! Just a few nits in the codeowner files.

Nik-09 committed:
Addressed review comments

Nik-09 committed:
CSS property updated

Nik-09 reviewed (commented):
Hi @kevintab95, I have addressed the review comments please take a look once you are free.

Thank you.

Nik-09 reviewed (commented):
Hi, @seanlip I have updated the error msg as per the discussion. Please provide your feedback on the updated video.

Thank you.

seanlip reviewed (approved):
Thanks @Nik-09, looks great!

Only comment I have is: for topics with no prerequisites, have an italicized gray "None" in the prerequisites column -- don't just leave it empty.

Thanks!

Nik-09 committed:
Adds frontend unit tests.

Nik-09 committed:
Adds backend unit test

Nik-09 committed:
UI update

Nik-09 committed:
Adds frontend unit test coverage.

kevintab95 reviewed (approved):
LGTM, thanks @Nik-09!

Nik-09 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into topic_dependency_graph

Nik-09 committed:
Merge branch 'develop' into topic_dependency_graph
